Make RandomSmallSemigroup random

pramothragavan opened the pull request:
There seems to be an issue where RandomSmallSemigroup gives the same semigroup each time. RandomSource defaults to a specific initial seed each time it's called, but removing it seems to make it work!

pramothragavan commented:
I'm not sure this change quite does what we want. While repeated calls to the function within a single GAP session yield different small semigroups, starting a new GAP session results in the same sequence of outputs each time (because we initialise with the same seed).

I think doing ``RandomSource(IsRealRandomSource, "random")`` instead of ``RandomSource(IsMersenneTwister)`` would work, but would also introduce an IO dependency. Is IO included by default?

Or maybe at the start of the function we could declare a random source ``r := RandomSource(IsMersenneTwister, Runtime())``, but I don't think that's quite random enough.

james-d-mitchell commented:
So, I had something of the same question as I think you've run into @pramothragavan. My understanding is that we don't really want "real" randomness by default, because this makes things like testing very hard to debug. An alternative solution would be to simply make a version of `RandomSmallSemi` which accepts a random source as the first argument. Then if you need "real" randomness, you can specify the source?

pramothragavan commented:
Sounds good, @james-d-mitchell! To clarify, you're saying this'd be a separate function (e.g. ``RandomSmallSemigroupWithSource``) and we'd keep the current changes to ``RandomSmallSemigroup``?

fingolfin commented:
I would suggest that you don't add a `RandomSmallSemigroupWithSource` function but rather follow the pattern set by other `RandomFOO` functions in GAP which tend to take an *optional* random source as first argument. So basically insert something like this:
```gap
  local rs, ...
  if Length(arg) >= 1 and IsRandomSource(arg[1]) then
    rs := Remove(arg, 1);
  else
    rs := GlobalMersenneTwister;
  fi;
```
Then pass `rs` to all calls to `Random` as first argument.

fingolfin commented:
Also to `RandomList` later in the function -- i.e. basically all `Random*` functions

pramothragavan commented:
Hi, sorry this has taken so long -- I've added in @fingolfin's suggestion now. I've changed the docs accordingly, but I'm not particularly happy about them (it becomes annoying to talk about odd/even arguments when you have an optional first argument) so any changes to make it more readable are very very welcome.

fingolfin reviewed (approved):
Looks pretty good to me now!

james-d-mitchell commented:
The CI isn't running, anyone have any ideas why?

wilfwilson reviewed (commented):
It would be good to add some tests of the new functionality.

pramothragavan commented:
I think CI should be fine now? Took a minute to construct an example that would enter the code block that codecov was complaining about

fingolfin commented:
I still looks fine to me but this is not my package, so @james-d-mitchell really should decide how to proceed :-)
